### PR TITLE
torch/_native: less per-call work in the chunked linear_cross_entropy loop

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #195830
* #197876
* __->__ #196134
* #195829

Three changes independent of the kernel's own arithmetic: `grad_linear_weight`
is no longer zeroed before the loop, because the first chunk writes it outright;
the kernel emits each row's loss contribution instead of the two statistics it
was built from, so the per-chunk loss arithmetic disappears; and the kernel's
staging default becomes `(512, 8)`. Together they are about 6% on top of the
kernel path on an H100 and about 10% on a B200, with every gradient bitwise
unchanged.

> None of the three changes what the loop computes. The first two remove work
> that existed only because of how the loop was written; the third changes only
> how the kernel schedules the work it already does.
>
> `grad_linear_weight` was allocated with `zeros` and then accumulated into,
> so the first chunk read back a (C, F) buffer that was known to be zero, after
> a (C, F) fill that nothing consumed. The first chunk now writes it with
> `mm(out=)` and later chunks accumulate as before, which drops a fixed
> `2 * C * F` bytes per call -- independent of the batch, the chunk size and the
> chunk count, so it reads as a share of call length: 4.8% at N=1024, 1.0% at
> N=16384. The accumulator is still zeroed when there is no first chunk to
> write it, which is the empty-batch early return.
>
> The kernel emitted `lse` and the target logit, and the caller then formed
> `scale * (lse - target_logit)` and reduced it, per chunk, on (Bc,) data:
> four launches moving 2 KB each. Both values are already in registers where
> the row's statistics are finished, so the kernel forms that product itself
> and writes one vector; the caller keeps one slot per row of the call and
> reduces once, after the loop, with `sum(out=loss)`. Nothing else read the two
> statistics, so they are gone from the kernel's signature. That slot is the one
> buffer here that is not a chunk, so the forward-only path -- which forms the
> loss chunk by chunk and never reads it -- does not allocate it.
>
> The staging default moves from `(512, 4)` to `(512, 8)` -- eight column tiles
> staged per barrier in pass 2 rather than four. `tiles_per_stage` does not
> reach pass 1's reduction, so this is a scheduling change and nothing else.
> Measured by calling the kernel directly from a captured CUDA graph, so that no
> GEMM time is in the number, over seven `(Bc, V)` chunk shapes -- `V` in
> {4096, 8192, 16384, 32000, 65536} at `Bc = 4096`, plus `Bc` in {1024, 2048} at
> `V = 32000` -- in both dtypes on an H100 and a B200, so 28 points: `(512, 8)`
> is faster at every one, by 1.4% to 12.2%, and uses less energy at every one,
> by 3.3% to 8.5%.
>
> Energy is worth reporting separately because it does not follow from the time.
> Per-call joules come from NVML's total-energy counter read over windows
> aligned to its updates, with the idle gaps subtracted at measured idle power.
> Where the board is not power-limited the faster setting draws 2-4% more power,
> so it saves less energy than time -- on B200 fp16, 9.1-12.2% faster for
> 6.9-8.5% less energy. Where the board is at its limit the two coincide: B200
> bf16 at `V >= 32000` runs both settings at ~1000 W, and there the energy win
> equals the time win. On H100 bf16 the faster setting also draws less power,
> 673 W against 691 W at `V=65536`, so it saves more energy (4.0%) than time
> (1.4%). Neither metric prefers `(512, 4)` at any point measured.
>
> That kernel is 4-25% of a chunked call and the rest is three cuBLAS GEMMs, so
> the retune is worth 0.1-2.9% end to end, largest where `in_features` is small
> relative to `num_classes`.
>
> Measured through the op's 12-shape sweep, paired per point across three
> repetitions -- the two states differ only in two Python files, so the arms are
> swapped in place at each point and a pair is measured seconds apart in one
> session.
>
> |                                    | H100 bf16 | H100 fp16 | B200 bf16 | B200 fp16 |
> | ---------------------------------- | --------- | --------- | --------- | --------- |
> | geomean                            | 1.058     | 1.065     | 1.095     | 1.099     |
> | min                                | 1.015     | 1.025     | 1.029     | 1.024     |
> | max                                | 1.143     | 1.160     | 1.377     | 1.390     |
> | above 1.0 in all three repetitions  | 11/12     | 12/12     | 12/12     | 12/12     |
>
> The maxima are at `in_features=1024` on the B200 and `V=4096` on the H100,
> where the fixed costs dominate; the minima are at `in_features=8192` and
> N=16384, where the surrounding GEMMs are largest. The single point not above
> 1.0 in every repetition is H100 bf16 at N=16384: 1.037 over a 0.992-1.042
> spread.
>
> Peak memory moves by at most the one new buffer, the (N,) fp32 loss terms:
> 0 to +48 kB measured, never more than `N * 4` bytes and sometimes nothing at
> all where the allocator reuses a freed block.
>
> The loss's summation order changes -- one reduction over N rather than a
> partial sum per chunk -- so the value is deterministic but not guaranteed
> bit-identical to a per-chunk accumulation. Measured against the previous path
> at four shapes spanning one to four chunks, in both dtypes, `grad_input`,
> `grad_linear_weight`, `grad_linear_bias` and the loss all came back bitwise
> identical.
>
> Test Plan:
>
> The kernel's isolation test asserts the loss term against a torch reference
> rather than against `lse` and the target logit separately. That pins the
> difference only: the row max cancels out of the term, and out of `g`, so what
> those assertions see of it is its effect on the row sum and not its value.
> Its other job, keeping `exp` in range, is pinned by
> `test_a_lone_high_column_pins_the_row_max`, where the row's maximum sits
> alone in one thread's column and a partial lost at either reduction stage
> takes the row out of fp32. The uniform-logit test became a stronger invariant
> in the process (`term == scale * log(V)`, whatever the logit value), and the
> aliased-buffer test still pins that the target logit is read before any write
> can occupy its bytes, since the term is now the only path by which that logit
> reaches an output.
>
> ```
> python test/python_native/test_lce_fused_grad_logits_kernel.py
> python test/python_native/test_linear_cross_entropy_override.py
> python test/test_nn.py -k linear_cross_entropy
> OPINFO_RESTRICT_TO_DSL=cutedsl python test/test_ops.py -k linear_cross_entropy
> lintrunner -m origin/main
> ```
>
> All pass on an H100 (sm_90) and a B200 (sm_100): 55 kernel tests, 31 override
> tests, 226 test_nn linear_cross_entropy tests, and the DSL-restricted test_ops
> run over its 5 OpInfo entries.

---
_🤖 Drafted by Claude Code (an AI agent) and reviewed & approved by pearu._

### CODE REVIEW

**Comment by pearu** [comment]
The block knobs were left tuned on H100 in #195829, and a comment there said re-tuning belonged here. This is that: the default moves from `(512, 4)` to `(512, 8)`, measured on both architectures rather than read off the earlier kernel map.

> **Why the earlier map was the wrong instrument.** The 3x5 grid recorded during #195829 reads `(512, 8)` as -4.5% at its worst point, which is why `(512, 4)` shipped. Following that map further made the op monotonically *worse* -- `(256,8)` 0.991, `(128,16)` 0.964, `(64,16)` 0.891, `(32,32)` 0.783 -- which is the signature of a map measuring something other than what the op runs. It was taken under auto chunking; the op runs matched-memory chunk shapes.
>
> **Measured directly instead.** `fused_grad_logits_into` called from a captured CUDA graph, so no GEMM time is in the number, over seven `(Bc, V)` chunk shapes -- `V` in {4096, 8192, 16384, 32000, 65536} at `Bc = 4096`, plus `Bc` in {1024, 2048} at `V = 32000` -- in both dtypes on an H100 and a B200. 28 points. `(512, 8)` is faster at **all 28** and lower energy at **all 28**:
>
> | | time | energy |
> | --- | --- | --- |
> | H100 fp16 | +5.3 ... +8.9% | +5.4 ... +7.2% |
> | H100 bf16 | +1.4 ... +3.9% | +3.3 ... +4.7% |
> | B200 fp16 | +9.1 ... +12.2% | +6.9 ... +8.5% |
> | B200 bf16 | +5.9 ... +8.1% | +5.5 ... +7.1% |
>
> No shape, architecture or dtype prefers `(512, 4)` at the kernel.
>
> **Energy was measured, not inferred, and it does not simply track time.** Per-call joules come from NVML's total-energy counter over windows aligned to its updates -- it advances only every ~330-450 ms, so a naive block reads zero -- with idle gaps subtracted at measured idle power. Where the board is not power-limited the faster setting draws 2-4% more power, so it saves less energy than time (B200 fp16: 9.1-12.2% faster for 6.9-8.5% less energy). At the 1000 W cap the two coincide (B200 bf16 at `V >= 32000`). On H100 bf16 the faster setting draws *less* power, 673 W against 691 W at `V=65536`, so it saves more energy (4.0%) than time (1.4%).
>
> **How much survives to the op, and why it is less than the kernel numbers suggest.** This kernel is 4-25% of a chunked call; the rest is three cuBLAS GEMMs. The knob is not confined to it, though: staging depth changes what the logits buffer leaves in L2 for the adjacent GEMMs. Timed directly, minimum of 40-60 runs in one process:
>
> | shape | chunks | kernel | three GEMMs | net |
> | --- | --- | --- | --- | --- |
> | B200 fp16 `V=8192` | 4096 | +11.3% | +0.25% | ~ +0.9% |
> | H100 bf16 `in_features=2048` | 2912 + 1184 | +3.96% | **-0.61%** | +0.13% |
> | H100 bf16 `N=8192` | 6368 + 1824 | +3.23% | **-0.31%** | +0.01% |
>
> On H100 the coupling runs against the kernel win and very nearly cancels it at these shapes; on B200 it is mildly favourable. So the op-level effect is what the paired sweeps below measure, not the kernel win times its share -- that product is an upper bound that ignores the coupling.
>
> **Op level, 12 shapes, three paired repetitions, matched memory.** Resolved effects only -- all three repetitions agreeing in sign:
>
> | | resolved wins | resolved losses |
> | --- | --- | --- |
> | H100 fp16 | 10/12 | 0 |
> | H100 bf16 | 6/12 | 2 |
> | B200 fp16 | 8/12 | 1 |
> | B200 bf16 | 9/12 | 0 |
>
> **The three losses were traced end to end** -- they are the rows in the coupling table above. The kernel is 3.2-11.3% faster at each, the GEMM coupling is -0.6% to +0.25%, and the net is between +0.01% and +0.9%. Nothing measured in the call produces a loss, so the sweep's -0.9% to -1.4% at those shapes is about a percentage point larger than anything the knob can account for, and is most likely a property of that measurement -- one process per arm -- rather than of the knob. Three repetitions agreeing in sign can resolve a systematic, not only an effect.
>
> **Other candidates rejected.** `(1024, 4)`: 7 resolved losses on B200 in each dtype. `(1024, 8)`: holds the single largest win seen anywhere (+5.6%, H100 fp16 at `V=65536`) but regresses at small vocabularies on H100 (-3.9%, -4.6%).

**Comment by slayton58** [comment]
@claude review this change, look closely for any unintended side-effects. Only surface blockers and major issues.

**Review by slayton58 — APPROVED** [review]
Looks good - didn't see anything concerning on a read through, or via claude - LGTM
